corehq/blobs/migrate.py

```diff
@@ -365,17 +387,89 @@ def get_document_provider(self):
 
 class BackendMigrator(Migrator):
 
+    has_worker_pool = True
+
     def __init__(self, slug):
         reindexer = BlobMetaReindexAccessor()
         types = [reindexer.model_class]
         assert not hasattr(types[0], "get_db"), types[0]  # not a couch model
         super(BackendMigrator, self).__init__(slug, types, BlobDbBackendMigrator)
         self.reindexer = reindexer
 
+    def get_doc_migrator(self, filename, **kw):
+        migrator = super(BackendMigrator, self).get_doc_migrator(filename)
+        return _migrator_with_worker_pool(migrator, self.reindexer, **kw)
+
     def get_document_provider(self):
         return SqlDocumentProvider(self.iteration_key, self.reindexer)
 
 
+@contextmanager
+def _migrator_with_worker_pool(migrator, reindexer, iterable, max_retry, num_workers):
+    """Migrate in parallel with worker pool
+
+    When running in steady state, failed doc will be retried up to the
+    max retry limit. Documents awaiting retry and all documents that
+    started the migration process but did not finish will be saved and
+    retried on the next run if the migration is stopped before it
+    completes.
+    """
+    def work_on(doc, key, retry_count):
+        try:
+            ok = migrator.migrate(doc)
+            assert ok, "run_with_worker_pool expects success!"
+        except Exception:
+            err = traceback.format_exc().strip()
+            print("Error processing blob:\n{}".format(err))
+            if retry_count < max_retry:
+                print("will retry {}".format(key))
+                retry_blobs[key] += 1
+                queue.put(doc)
+                return
+            migrator.save_backup(doc, "too many retries")
+            print("too many retries {}".format(key))
+        retry_blobs.pop(key, None)
+
+    def retry_loop():
+        for doc in queue:
+            enqueue_doc(doc)
+
+    def enqueue_doc(doc):
+        key = reindexer.get_key(doc)
+        retry_count = retry_blobs.setdefault(key, 0)
+        # pool.spawn will block until a worker is available
+        pool.spawn(work_on, doc, key, retry_count)
+        # Returning True here means the underlying iterator will think
+        # this doc has been processed successfully. Therefore we must
+        # process this doc before the process exits or save it to be
+        # processed on the next run.
+        return True
+
+    queue = LifoQueue()
```

Comment:
Why `lifo`?

Reply:
So the retry loop does not process all items in the queue before stopping (on keyboard interrupt, etc.). Another option is to use a different signal/event channel and block on the queue with a timeout—checking for the stop event periodically. More complicated, probably doesn't matter that much.